sim/firesim-lib/src/main/scala/bridges/DromajoBridge.scala

```diff
@@ -0,0 +1,178 @@
+    lazy val TOKEN_WIDTH = 512 // in b
+    lazy val toHostCPUQueueDepth = QUEUE_DEPTH
+    lazy val dmaSize = BigInt((TOKEN_WIDTH / 8) * QUEUE_DEPTH)
+    // CONSTANTS
+
+    // setup io
+    val io = IO(new WidgetIO)
+    val hPort = IO(HostPort(new DromajoTargetIO(key.insnWidths, key.vecSizes)))
+
+    // the target is ready to both send/receive data
+    val tFire = hPort.toHost.hValid && hPort.fromHost.hReady
+
+    // helper to get number to round up to nearest multiple
+    def roundUp(num: Int, mult: Int): Int = { (((num - 1) / mult) + 1) * mult }
+
+    // get the traces
+    val traces = hPort.hBits.trace.insns.map({ unmasked =>
+      val masked = WireDefault(unmasked)
+      masked.valid := unmasked.valid && !hPort.hBits.trace.reset
+      masked
+    })
+    private val iaddrWidth = roundUp(traces.map(_.iaddr.getWidth).max, 8)
+    private val insnWidth  = roundUp(traces.map(_.insn.getWidth).max, 8)
+    private val wdataWidth = roundUp(traces.map(_.wdata.getWidth).max, 8)
+    private val causeWidth = roundUp(traces.map(_.cause.getWidth).max, 8)
+    private val tvalWidth  = roundUp(traces.map(_.tval.getWidth).max, 8)
+
+    // hack since for some reason padding a bool doesn't work...
+    def boolPad(in: Bool, size: Int): UInt = {
+      val temp = Wire(UInt(size.W))
+      temp := in.asUInt
+      temp
+    }
+
+    val paddedTraces = traces.map { trace =>
+      Cat(trace.tval.pad(tvalWidth),
+        Cat(trace.cause.pad(causeWidth),
+          Cat(boolPad(trace.interrupt, 8),
+            Cat(boolPad(trace.exception, 8),
+              Cat(trace.priv.asUInt.pad(8),
+                Cat(trace.wdata.pad(wdataWidth),
+                  Cat(trace.insn.pad(insnWidth),
+                    Cat(trace.iaddr.pad(iaddrWidth), boolPad(trace.valid, 8)))))))))
+    }
+
+    val maxTraceSize = paddedTraces.map(t => t.getWidth).max
+    val outDataSzBits = outgoingPCISdat.io.enq.bits.getWidth
+
+    // constant
+    val totalStreamsPerToken = 2 // minTraceSz==190b so round up to nearest is 256b
+                                 // constant
+
+    require(maxTraceSize < (outDataSzBits / totalStreamsPerToken), "All instruction trace bits must fit in 256b")
+
+    // how many streams being sent over
+    val numStreams = traces.size
+    // num tokens needed to display full stream
+    val numTokenForAll = ((numStreams - 1) / totalStreamsPerToken) + 1
+
+    // tracequeue is full
+    val traceQueueFull = outgoingPCISdat.io.deq.valid && !outgoingPCISdat.io.enq.ready
+
+    // only inc the counter when the something is sent... input is valid... and output is avail
+    val counterFire = outgoingPCISdat.io.enq.fire() && hPort.toHost.hValid && !traceQueueFull
+    val (cnt, wrap) = Counter(counterFire, numTokenForAll)
+
+    val paddedTracesAligned = paddedTraces.map(t => t.asUInt.pad(outDataSzBits/totalStreamsPerToken))
+    val paddedTracesTruncated = if (numStreams == 1) {
+      (paddedTracesAligned.asUInt >> (outDataSzBits.U * cnt))
```

Comment:
I'm not sure if a dynamic shift is what you want to do here, instead of say making a Vec and using dynamic indexing.  Vivado might not care though.

Reply:
Vivado didn't seem to care when testing Jerry's SPEC stuff.

Reply:
By caring i'm referring to resource utilization, and by not care i'm referring to doing a strength reduction on the dynamic shift.

Reply:
Up to you if you want me to change this. IMO this doesn't matter much because we ran Fromajo on MegaBOOM which is currently the largest design that you might want to run Fromajo with (and there didn't seem to be resource issues).

Reply:
I'm fine with it.